### PR TITLE
Adding an example on  Compact Convolutional Transformers

### PR DESCRIPTION
**Pull request by sayakpaul**
Colab Notebook is available [here](https://colab.research.google.com/drive/1McaDiYultDUaJ5psNLpGHq_yE8P5n7yn?usp=sharing).

### CODE REVIEW

**Comment by alihassanijr** [comment]
Hi,
I do want to add this example is CCT-2/3x2, which is our smallest model and should be at around 0.28M params, not 0.4M which is noted in the notebook (by taking a quick look I noticed the MHSA from Keras accounts for a lot of the increased parameters, so maybe there's something there worth exploring?). There are some details in the training that help achieve the best results including but not limited to cosine LR schedule and more augmentation policies (AutoAug or Mixup and CutMix). However at this level the results are more or less the same with what we ran ourselves so it shouldn’t matter much.

**Comment by sayakpaul** [comment]
Thanks for sharing your comments. I have included the notes on augmentation already inside the notebook. Will add some more notes on the model being CCT-2/3x2 as well as comments on parameters. 

For the slight increase in parameters, I'd not worry too much since 0.4 M and 0.28 M are not that drastic.

**Comment by alihassanijr** [comment]
Thank you.
Our model is about 30% smaller. That difference will only grow as larger models are used, which are needed for better performance. Our best model on CIFAR-10 is 3.17M parameters. So if this scales it would be over 10M (almost as big as a ResNet18). A key aspect of our model is it’s compactness, or how it achieves high performance with a low number of parameters.
And I can confirm that when I rewrote the MHSA myself, model.summary() returned 275K params.

**Comment by sayakpaul** [comment]
Yeah got it. But the purpose of [Keras Examples](https://keras.io/examples) is also to showcase the use of how we can use the factory layers to put an idea to use quickly with Keras best practices incorporated. I can note all your points down inside the example to provide a more holistic view to the readers. So, reimplementing MHSA in this example is a bit out of scope. I hope you'd understand.

> A key aspect of our model is it’s compactness, or how it achieves high performance with a low number of parameters.

IMO, the current example does convey this idea (I noted the performance and params with respect to a vanilla ViT) but I understand your point on making it even more compact.

**Comment by alihassanijr** [comment]
That makes sense. We appreciate you integrating our work!

**Review by fchollet — COMMENTED** [review]
Excellent work!

Note that I've pushed some minor copyedits, please pull them first.

File: examples/vision/cct.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+    augmented = data_augmentation(inputs)
+
+    # Encode patches.
+    cct_tokenizer = CCTTokenizer()
+    encoded_patches = cct_tokenizer(augmented)
+
+    # Apply positional embedding.
+    if positional_emb:
+        pos_embed, seq_length = cct_tokenizer.positional_embedding(image_size)
+        positions = tf.range(start=0, limit=seq_length, delta=1)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/cct.py:251]: `seq_length` is being calculated from the image shape and the kernel size provided. Apart from that, I'll make the rest of the things to be accepted as arguments.

**Comment by sayakpaul** [comment]
@fchollet addressed all the feedback you provided. Thank you.

**Review by fchollet — APPROVED** [review]
Awesome! Please add the generated files.

**Comment by sayakpaul** [comment]
@fchollet done. Ready for you to take over.

**Comment by fchollet** [comment]
Thank you for another great contribution! 👍

**Pull request merged**